pkg/agent/controller/networkpolicy/l7engine/l7visbility_watcher.go

```diff
@@ -0,0 +1,397 @@
+	_, err := trafficControlApi.Create(context.TODO(), tc, metav1.CreateOptions{})
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
+		klog.ErrorS(err, "Failed to create TrafficControl")
+		return err
+	}
+	return nil
+}
+
+func (l7w *l7Watcher) postTrafficControl(tcName string, direction v1alpha2.Direction, podLabels, namespaceLabels map[string]string) error {
+	targetPort := &v1alpha2.OVSInternalPort{Name: config.L7NetworkPolicyTargetPortName}
+	var err error
+	if len(podLabels) != 0 {
+		err = createTrafficControl(tcName, namespaceLabels, podLabels, direction, v1alpha2.ActionMirror, targetPort, l7w.tcController)
+	} else {
+		err = createTrafficControl(tcName, namespaceLabels, nil, direction, v1alpha2.ActionMirror, targetPort, l7w.tcController)
+	}
+	if err != nil {
+		klog.ErrorS(err, "Error Creating Traffic Control")
+		return err
+	}
+	err = l7w.tcController.SyncTrafficControl(tcName)
+	if err != nil {
+		klog.ErrorS(err, "Error Syncing Traffic Control")
+		return err
+	}
+	return nil
+}
+
+func (l7w *l7Watcher) deleteTrafficControl(tcName string) error {
+	return l7w.tcController.CrdClient.CrdV1alpha2().TrafficControls().Delete(context.TODO(), tcName, metav1.DeleteOptions{})
+}
+
+func generateTCName(direction v1alpha2.Direction, podLabels, namespaceLabels map[string]string) (string, error) {
+	directionStr := "Both"
+	switch direction {
+	case v1alpha2.DirectionIngress:
+		directionStr = "Ingress"
+	case v1alpha2.DirectionEgress:
+		directionStr = "Egress"
+	case v1alpha2.DirectionBoth:
+		directionStr = "Both"
+	}
+	if podLabels == nil {
+		podLabels = make(map[string]string)
+	}
+	data := map[string]interface{}{
+		"direction":       directionStr,
+		"podLabels":       podLabels,
+		"namespaceLabels": namespaceLabels,
+	}
+	dataJSON, err := json.Marshal(data)
+	if err != nil {
+		klog.ErrorS(err, "Error creating hash for Traffic Control")
+		return "", err
+	}
+	hashval := sha256.Sum256(dataJSON)
+	hash := fmt.Sprintf("%x", hashval[:14])
+	return hash, nil
+}
+
+func (l7w *l7Watcher) addToLabelsToTCMap(labelsKey, tcName string) {
+	if tcArray, ok := l7w.labelsToTCMap[labelsKey]; ok {
+		l7w.labelsToTCMap[labelsKey] = append(tcArray, tcName)
+	} else {
+		l7w.labelsToTCMap[labelsKey] = []string{tcName}
+	}
+}
+
+func createLabelsKey(podLabels, namespaceLabels map[string]string) (string, error) {
+	data := map[string]interface{}{
+		"podLabels":       podLabels,
+		"namespaceLabels": namespaceLabels,
+	}
+	dataJSON, err := json.Marshal(data)
+	if err != nil {
+		klog.ErrorS(err, "Error creating hash for Traffic Control")
+		return "", err
+	}
+	hashval := sha256.Sum256(dataJSON)
+	key := fmt.Sprintf("%x", hashval[:14])
+	return key, nil
+}
+
+func (l7w *l7Watcher) startSuricataAndCreateTC(direction v1alpha2.Direction, nodeName string, podLabels, namespaceLabels map[string]string) error {
+	err := l7w.StartSuricataIfNotRunning(nodeName)
+	if err != nil {
+		klog.Error("Suricata not started!")
```

Comment:
should be `klog.ErrorS` and include the error. You may want to review all the logs which are part of this PR

Reply:
Not required anymore